### PR TITLE
Initial change to compile middleware invoke method when extra args 

### PR DESCRIPTION
**Pull request by davidfowl**
#530 

We could maybe move some of this compilation logic to Common (not sure if we do it in more than one place).

/cc @DamianEdwards @pranavkm

### CODE REVIEW

File: test/Microsoft.AspNet.Http.Abstractions.Tests/UseMiddlewareTest.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using Microsoft.AspNet.Builder;
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Http.Abstractions.Tests/UseMiddlewareTest.cs]: @mikaelm12 didn't format this file properly :disappointed:

**Comment by JamesNK** [comment]
Do you need to worry about methods that could have ref or out parameters? You should probably either add support for them or throw an exception with an error message when building the expression tree.

**Comment by davidfowl** [comment]
@JamesNK seems unrelated to this change in particular. What happens when you do methodInfo.Invoke and you have ref and out params? Does it just work?

**Comment by JamesNK** [comment]
Yeah it does so this would be a regression.

This is a method I wrote to handle calling pretty much any method info via a compiled expression: https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Utilities/ExpressionReflectionDelegateFactory.cs#L86-L190

Looking at what I wrote, you might be ok with ref parameters because you don't need to capture the return value. You probably shouldn't try to resolve out params from DI. Either leave them alone or tell the user it is an error to have out parameter here.

File: src/Microsoft.AspNet.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs [hunk 2]
```diff
@@ -80,20 +88,97 @@ public static IApplicationBuilder UseMiddleware(this IApplicationBuilder app, Ty
...
+            methodArguments[0] = httpContextArg;
+            for (int i = 1; i < parameters.Length; i++)
+            {
+                var parameterType = parameters[i].ParameterType;
+                if (parameterType.IsByRef)
+                {
+                    throw new NotSupportedException(Resources.FormatException_InvokeDoesNotSupportRefOrOutParams(InvokeMethodName));
+                }
+
+                var parameterTypeExpression = new Expression[] {
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs]: `c#
new Expression[]
{

};
```
(brace on separate line)

**Comment by pranavkm** [comment]
:shipit:

**Pull request merged**